contributivity.py

```diff
@@ -3,32 +3,88 @@
+from math import factorial
+from sklearn.linear_model import LinearRegression
+from scipy.special import softmax
+
+import fl_training
+import shapley_value.shapley as sv
+
+
+class krigingModel:
+    def __init__(self, degre, covariance_func):
+        self.X = np.array([[]])
+        self.Y = np.array([[]])
+        self.cov_f = covariance_func
+        self.degre = degre
+        self.beta = np.array([[]])
+        self.H = np.array([[]])
+        self.K = np.array([[]])
+        self.invK = np.array([[]])
+
+    def fit(self, X, Y):
+        self.X = X
+        self.Y = Y
+        K = np.zeros((len(X), len(X)))
+        H = np.zeros((len(X), self.degre + 1))
+        for i, d in enumerate(X):
+            for j, b in enumerate(X):
+                K[i, j] = self.cov_f(d, b)
+            for j in range(self.degre + 1):
+                H[i, j] = np.sum(d) ** j
+        self.H = H
+        self.K = np.linalg.inv(K)
+        self.invK = np.linalg.inv(K)
+        Ht_invK_H = H.transpose().dot(self.invK).dot(H)
+        self.beta = (
+            np.linalg.inv(Ht_invK_H).dot(H.transpose()).dot(self.invK).dot(self.Y)
+        )
+
+    def predict(self, x):
+        gx = []
+        for i in range(self.degre + 1):
+            gx.append(np.sum(x) ** i)
+        gx = np.array(gx)
+        cx = []
+        for i in range(len(self.X)):
+            cx.append([self.cov_f(self.X[i], x)])
+        cx = np.array(cx)
+        pred = gx.transpose().dot(self.beta) + cx.transpose().dot(self.invK).dot(
+            self.Y - self.H.dot(self.beta)
+        )
+        return pred
 
 
 class Contributivity:
-    def __init__(
-        self,
-        name="",
-        contributivity_scores=np.array([]),
-        scores_std=np.array([]),
-        computation_time=0,
-    ):
+    def __init__(self, name="", scenario=None):
         self.name = name
-        self.contributivity_scores = contributivity_scores
-        self.scores_std = scores_std
-        sum_contrib = np.sum(contributivity_scores)
-        if len(contributivity_scores) == 0:
-            self.normalized_scores = np.array([])
-        else:
-            self.normalized_scores = contributivity_scores / sum_contrib
-        self.computation_time = computation_time
+        n = len(scenario.node_list)
+        self.contributivity_scores = np.zeros(n)
+        self.scores_std = np.zeros(n)
+        self.normalized_scores = np.zeros(n)
+        self.computation_time = 0.0
+        self.charac_fct_calls_count = 0
+        self.charac_fct_values = {(): 0}
+        self.increments_values = []
+        for i in range(n):
+            self.increments_values.append(dict())
 
     def __str__(self):
         computation_time = str(datetime.timedelta(seconds=self.computation_time))
         output = "\n" + self.name + "\n"
         output += "Computation time: " + computation_time + "\n"
+        output += (
+            "Number of characteristic function computed: "
+            + str(self.charac_fct_calls_count)
```

Comment:
You can save this number in the csv results file here --> https://github.com/SubstraFoundation/distributed-learning-contributivity/blob/13b25a2aa0397410cf67416a84df099326d3e66e/scenario.py#L373